### ISSUE TITLE
Add class Map for array and Matrix(Symbol)

### ISSUE THREAD

**Issue by stevenleeS0ht** [sympy/sympy#18431]
I don't know if sympy support SIMD feature for Sympy collections operation.

The collection here include Array, Matrix, Matrix Symbol.

for example like this:
```python3
>>> M=Matrix([1,2,3])
>>> Map(M, Lambda(x, sin(x)))
⎡sin(1)⎤
⎢      ⎥
⎢sin(2)⎥
⎢      ⎥
⎣sin(3)⎦
```

If it is MatrixSymbol, it would become like that:

```python3
>>> M=MatrixSymbol('M')
>>> Map(M,Lambda(y,cos(y)))
Map(M, y ↦ cos(y))
```

**Comment by gschintgen** [comment]
For matrices at least there's `applyfunc`: https://docs.sympy.org/1.5.1/modules/matrices/common.html?highlight=applyfunc#sympy.matrices.common.MatrixCommon.applyfunc

**Comment by sylee957** [comment]
There is also a fully symbolic node `ElementwiseApplyFunc` for matrix expressions, so I don't think that it's an issue for matrix.
But I won't close this until I hear `Array` have similar symbolic feature.

**Comment by Saanidhyavats** [comment]
I am interested to take up this issue. How should I start?

**Comment by sylee957** [comment]
I'd add a new class like `ArrayElementwiseApplyFunc` if it is missing. But I'm not familiar with the tensor module and whether it is really needed. But I can say that it's not needed for `Matrix`.

**Comment by Saanidhyavats** [comment]
What exactly I have to implement @sylee957 can you elaborate?

**Comment by Saanidhyavats** [comment]
@sylee957 can I add this class `ArrayElemenrwiseApplyFunc`  so that Map function can work on it.

**Comment by sylee957** [comment]
Yes, you should create a new class under tensor/array module
I think that it should be generalized for more than matrices.

**Comment by sylee957** [comment]
But I'm still unsure about whether every stuff of `matrices/expressions` should be mirrored over `tensor/arrays` is a good idea.
I'm also thinking of extending some operators of `matrices/expressions` for n-dimensional arrays and managing such things in one place.
I think that @Upabjojr should comment about this.

**Comment by Saanidhyavats** [comment]
Shall I implement it on `tensor/arrays/ndim_array.py` @sylee957

**Comment by sylee957** [comment]
No, I think that there is no need of method because Array.applyfunc already does that.
There can be a separate need for lazy-evaluation class instead for stuff like lazy evaluations, or code printing.

**Comment by Saanidhyavats** [comment]
> Yes, you should create a new class under tensor/array module
> I think that it should be generalized for more than matrices.

We already have a method `applyfunc` for n-dim array.  The class which we will be creating shall be generalised for matrices and arrays. Whether I am thinking in the right direction or there has to be something more in this.

**Comment by Upabjojr** [comment]
There are methods for `tensor/array` in `sympy/codegen/array_utils.py`.
